libpromises/expand.c

```diff
@@ -665,6 +667,11 @@ void BundleResolve(EvalContext *ctx, const Bundle *bundle)
     {
         BundleResolvePromiseType(ctx, bundle, "classes", VerifyClassPromise);
     }
+
+    /* Necessary to also parse vars *after* classes,
+     * because "inputs" might be affected. */
+    BundleResolvePromiseType(ctx, bundle, "vars",
+                             (PromiseActuator *) VerifyVarPromise);
```

Comment:
@nickanderson: here is the vars-classes-vars pre-eval passes. It used to be vars-classes, I changed it to classes-vars, but vars-classes-vars passes all tests, even several old soft-fails.

Overall pre-eval is a huge mess and we should rethink all of it. TODO ponder. ;-)